internal/services/service.go

```diff
@@ -45,33 +46,57 @@ func (s simpleService) Start(ctx context.Context) error {
 		<-ctx.Done()
 		logger.Debug.Printf("%v: %v\n", s.String(), ctx.Err())
 		if err := cmd.Process.Signal(syscall.SIGTERM); err != nil {
-			msg := "%v: failed to send SIGTERM to %v\n"
-			logger.Error.Printf(msg, s.String(), s.binaryName)
+			msg := "failed to send SIGTERM to %v: %v\n"
+			logger.Warning.Printf(msg, s.name, err)
 		}
 	}()
 
 	err := cmd.Wait()
-	if err != nil && cmd.ProcessState.ExitCode() != int(syscall.SIGTERM) {
+	exitCode := cmd.ProcessState.ExitCode()
+	signal := cmd.ProcessState.Sys().(syscall.WaitStatus).Signal()
+	if err != nil && exitCode != 0 && signal != syscall.SIGTERM {
 		return err
 	}
 	return nil
 }
 
-func (s simpleService) String() string {
-	return s.serviceName
+// Ready blocks until the service is ready or the context is canceled.
+//
+// A service is considered ready when it is possible to establish a connection
+// to its healthcheck endpoint.
+func (s Service) Ready(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			conn, err := net.Dial("tcp", fmt.Sprintf("0.0.0.0:%s", s.healthcheckPort))
+			if err == nil {
+				logger.Debug.Printf("%s is ready\n", s.name)
+				conn.Close()
+				return nil
+			}
+			time.Sleep(DefaultDialInterval)
+		}
+	}
+}
+
+func (s Service) String() string {
+	return s.name
 }
 
 // The Run function serves as a very simple supervisor: it will start all the
 // services provided to it and will run until the first of them finishes. Next
 // it will try to stop the remaining services or timeout if they take too long
-func Run(services []Service) {
+func Run(ctx context.Context, services []Service) {
 	if len(services) == 0 {
 		logger.Error.Panic("there are no services to run")
 	}
 
 	// start services
-	ctx, cancel := context.WithCancel(context.Background())
-	exit := make(chan struct{})
+	startedServicesCount := 0
```

Comment:
We can replace part of this logic with a WaitGroup.
https://pkg.go.dev/sync#example-WaitGroup

I will write a PR on top of your proposing something.